File: Math/Hamming_distance.py
'''
    The Hamming distance between two integers is the number of positions at which the corresponding bits are different.

    Given two integers x and y, return the Hamming distance between them.

    Example 1:
    Input: x = 1, y = 4
    Output: 2
    Explanation:
    1   (0 0 0 1)
    4   (0 1 0 0)
        ↑   ↑
    The above arrows point to positions where the corresponding bits are different.
    
    Example 2:
    Input: x = 3, y = 1
    Output: 1

    Constraints:
    0 <= x, y <= 231 - 1
'''
class Solution:
    def totalHammingDistance(self, nums: List[int]) -> int:
        '''By using bit manipulation, as all the array elements are 32-bit array elements, we calculate all the number of set bits and unset bits as we need to consider the permutations, we take setbits*unsetbits.'''
        hamming_dist,n = 0,len(nums)
        for i in range(32):
            count = 0
            for element in nums:
                #Right shifting the element by the index and performing &1 lets us know if a bit is set or not
                if((element>>i)&1):
                    count+=1
            #Adding all the combinations where there are set and unset bits.
            hamming_dist+=count*(n-count)
        return hamming_dist

File: Math/Hammingdistance.cpp
#include <vector>

class Solution {
public:
    int totalHammingDistance(std::vector<int>& nums) {
        // Get the number of elements in the input vector
        int n = nums.size();
        // Initialize the variable to store the total Hamming distance
        int ans = 0;
        
        // Iterate through each bit position (from 0 to 31)
        for (int i = 0; i < 32; i++) {
            // Initialize a variable to count the number of set bits at the current bit position
            int count = 0;
            
            // Iterate through all elements in the vector
            for (int k = 0; k < n; k++) {
                // Count the number of set bits at the current bit position for each element
                count += (nums[k] >> i) & 1;
            }
            
            // Update the total Hamming distance by adding the product of set bits and unset bits
            ans += count * (n - count);
        }
        
        // Return the total Hamming distance
        return ans;
    }
};

/*
This program defines a Solution class with a totalHammingDistance method that calculates the total Hamming distance for the input vector nums. The outer loop iterates over each bit position (from 0 to 31), and the inner loop counts the number of set bits at that bit position for all elements in the vector. It then updates the ans variable by adding the product of the count of set bits and the count of unset bits at that position.

This code efficiently calculates the total Hamming distance for a vector of integers by considering each bit position separately. */
